src/audio/fft/fft.c

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Amery Song <chao.song@intel.com>
+//	   Keyon Jie <yang.jie@linux.intel.com>
+
+#include <sof/common.h>
+#include "fft.h"
+#include "twiddle.h"
+
+static void icomplex32_add(struct icomplex32 *in1, struct icomplex32 *in2, struct icomplex32 *out)
+{
+	out->real = in1->real + in2->real;
+	out->image = in1->image + in2->image;
+}
+
+static void icomplex32_sub(struct icomplex32 *in1, struct icomplex32 *in2, struct icomplex32 *out)
+{
+	out->real = in1->real - in2->real;
+	out->image = in1->image - in2->image;
+}
+
+static void icomplex32_mul(struct icomplex32 *in1, struct icomplex32 *in2, struct icomplex32 *out)
+{
+	int64_t tmp;
+
+	tmp = (int64_t)in1->real * (int64_t)in2->real - (int64_t)in1->image * (int64_t)in2->image;
+	out->real = (int)(tmp >> 31);
+	tmp = (int64_t)in1->real * (int64_t)in2->image + (int64_t)in1->image * (int64_t)in2->real;
+	out->image = (int)(tmp >> 31);
+}
+
+void fft(struct icomplex32 *inb, struct icomplex32 *outb, unsigned int size)
+{
+	struct icomplex32 tmp1, tmp2;
+	int bitReverseIdx[4096] = {0};
+	int lim = 1, len = 0;
+	int i, j, k, m;
+	int depth;
+
+	while (lim < size) {
+		lim <<= 1;
+		len++;
+	}
+
+	for (i = 1; i < lim; ++i) {
+		bitReverseIdx[i] = (bitReverseIdx[i >> 1] >> 1) | ((i & 1) << (len - 1));
+		outb[bitReverseIdx[i]] = inb[i];
+	}
+
+	for (depth = 1; depth <= len; ++depth) {
+		m = 1 << depth;
+
+		for (k = 0; k < lim; k += m) {
+			for (j = 0; j < m / 2; ++j) {
+				tmp1.real = twiddle_real[j * (1024 / m)];
```

Comment:
Move compute of 1024 / m outside looops for k and j.

Reply:
> Move compute of 1024 / m outside looops for k and j.

good, this will help for performance!

Reply:
m is also a power of 2 afaict so please shift.